### DISCUSSION TITLE
Out of memory error

### DISCUSSION THREAD

**Discussion by alexandrebtessael** [GEOS-DEV/GEOS#1903] [Q&A] [answered]
Hello ! I am having some memory issues while performing poromechanics simulation on an imported full hex grid (see screenshot below). However, there is no problem with the mesh because I used it for compositionalMultiphaseFlow and geomechanics simulations and this error did not happen. I also performed poromechanics simulations on grids with less cells but with multiple elements and the error did not happen.

We did some debugging but that did not help. Any idea of the cause of this error ?

![image](https://user-images.githubusercontent.com/87317386/166905332-ba98c0ec-4832-46e8-9613-66f537dc8365.png)

**Comment by francoishamon** [comment]
Hello @alexandrebtessael,

If your problem is too large, the direct linear solver is likely to run out of memory. I recommend using:
```
      <LinearSolverParameters                                                                                                                                                            
        solverType="gmres"                                                                                                                                                               
        preconditionerType="mgr"                                                                                                                                                         
        krylovTol="1.0e-6"                                                                                                                                                                                                                                                                                                                
        logLevel="1"/>
```
to switch to an iterative linear solver. You can choose `krylovTol` to be one order of magnitude smaller than your `newtonTol`. With the `logLevel="1"`, GEOSX will output the number of linear iterations and some other statistics. If you see that this number is above 100, try:
```
      <LinearSolverParameters                                                                                                                                                            
        solverType="gmres"                                                                                                                                                               
        preconditionerType="mgr"                                                                                                                                                         
        krylovTol="1.0e-6"                                                                                                                                                               
        krylovAdaptiveTol="1"                                                                                                                                                            
        krylovWeakestTol="0.1"                                                                                                                                                           
        logLevel="1"/>
```
and if the linear solver still struggles, please post the xml and the simulation log

**Comment by alexandrebtessael** [comment]
Thanks for answering @francoishamon , I tried your method and it worked. I don't encounter this memory issue anymore.

I have another question linked to the LinearSolverParameters when it comes to CO2 injection. I used the direct linear solver because when I use : 
<LinearSolverParameters                                                                                                                                                            
        solverType="gmres"                                                                                                                                                               
        preconditionerType="mgr"

I encounter the "Error!! Block ordering for non-Galerkin coarse grid is not currently supported".  But with large problems, I face the memory issue with direct linear solver.

Is there an another solverType or preconditionerType that I could use to avoid this ?

**Reply by francoishamon** [reply]
Hello @alexandrebtessael,

Sorry about the issue that you are facing with wells, this is a known issue of the linear solver that will hopefully be fixed next week. In the meantime, you can replace the wells with boundary conditions to overcome the problem. There is an example here: https://github.com/GEOSX/GEOSX/blob/develop/inputFiles/compositionalMultiphaseFlow/deadoil_3ph_staircase_3d.xml

Best,

Francois

**Comment by alexandrebtessael** [comment]
Hello @francoishamon , thanks for the answer, I will use this solution for the moment.

**Comment by francoishamon** [comment, marked as answer]
@alexandrebtessael If you want to try again the problem with wells, the iterative linear solver with:
```
<LinearSolverParameters
solverType="gmres"
preconditionerType="mgr"
```
should work now with the latest version of `develop`. Let me know if you have some issues
